lib/burnside/intrinsics.cc

```diff
@@ -334,26 +441,35 @@ static mlir::FunctionType getFunctionType(mlir::Type resultType,
       argumentTypes, resultType, getModule(&builder).getContext());
 }
 
+// TODO find nicer type to string infra or move this in a mangling utility
+// mlir as Type::dump(ostream) methods but it may adds !
+static std::string typeToString(mlir::Type t) {
+  if (auto i{t.dyn_cast<mlir::IntegerType>()}) {
+    return "i" + std::to_string(i.getWidth());
+  } else if (auto cplx{t.dyn_cast<fir::CplxType>()}) {
+    return "z" + std::to_string(cplx.getFKind());
+  } else if (auto real{t.dyn_cast<fir::RealType>()}) {
+    return "r" + std::to_string(real.getFKind());
+  } else if (auto f{t.dyn_cast<mlir::FloatType>()}) {
+    return "f" + std::to_string(f.getWidth());
+  } else if (auto logical{t.dyn_cast<fir::LogicalType>()}) {
+    return "l" + std::to_string(logical.getFKind());
+  } else if (auto character{t.dyn_cast<fir::CharacterType>()}) {
+    return "c" + std::to_string(character.getFKind());
+  } else {
```

Comment:
Hmm. All types know how to print themselves in MLIR and FIR. I wonder if that can be leveraged somehow.

Reply:
Maybe, I looked a bit, but the best I could find is `Type::print(ostream)` that is based on `ModulePrinter::printType(Type)`: https://github.com/tensorflow/mlir/blob/e81c6f9c15a454539a0b41bfbbe99441ea33c840/lib/IR/AsmPrinter.cpp#L984.
I am not sure the type printed by this can be used in function names. I think `fir.acos.!fir.real<16>` is not a valid `symbol-ref-id` (https://github.com/tensorflow/mlir/blob/master/g3doc/LangRef.md#identifiers-and-keywords), but maybe it actually is. Would you be OK having such intrinsic function names if it is legal ? If so, I can try using  `Type::print(ostream)`.

Reply:
So after reading mlir doc a bit more and testing, it seems to be legal to use `fir.acos.!fir.real<16>` an mlir function name (so we can use `Type::print(llvm::raw_ostream)`). The issue is more if these function names go until code generation; I am not sure it would be legal to have such names in all target assemblies. On X86, it surprisingly seems to compile, link and run OK. So `!` and `<` seems just fine in x86 elf symbol names but I am not sure we should do it...

What's your take, do you see a way we could use `Type::print(llvm::raw_ostream)` ?

Reply:
Thanks for experimenting with this. We can leave it in our back of tricks at this point.

The concern was more about proliferation of different mangling and printing schemes in the bridge(s).